shared/manifest.go

```diff
@@ -68,27 +71,87 @@ func (m *Manifest) unmarshalAll() error {
 	return nil
 }
 
-// Contains checks whether m contains the path.
-func (m *Manifest) Contains(path string) (bool, error) {
-	if m.imap == nil {
-		if err := m.unmarshalAll(); err != nil {
-			return false, err
-		}
+func findNode(t interface{}, parts []string) interface{} {
+	if len(parts) == 0 {
+		return t
 	}
-	parts := strings.Split(path, "/")
-	// Split always returns at least one element.
-	// Remove the leading empty part.
-	if parts[0] == "" {
-		parts = parts[1:]
+
+	// t could be nil (e.g. if the previous part does not exist in the map), in which case casting will fail.
+	trie, ok := t.(map[string]interface{})
+	if !ok {
+		return nil
 	}
+	return findNode(trie[parts[0]], parts[1:])
+}
+
+// ContainsFile checks whether m contains a file path (including directories).
+func (m *Manifest) ContainsFile(path string) (bool, error) {
+	if err := m.unmarshalAll(); err != nil {
+		return false, err
+	}
+
+	path = strings.Trim(path, "/")
+	if path == "" {
+		// Root directory always exists.
+		return true, nil
+	}
+	parts := strings.Split(path, "/")
 	for _, items := range m.imap {
-		if trieContains(items, parts) {
+		if findNode(items, parts) != nil {
 			return true, nil
 		}
 	}
 	return false, nil
 }
 
+// ContainsTest checks whether m contains a full test URL.
+func (m *Manifest) ContainsTest(testURL string) (bool, error) {
+	if err := m.unmarshalAll(); err != nil {
+		return false, err
+	}
+
+	// URLs in the manifest do not include the leading slash (url_base).
+	testURL = strings.TrimLeft(testURL, "/")
+	path, query := ParseTestURL(testURL)
+	parts := strings.Split(path, "/")
+	// parts=["foo", "bar", "test.any.js"]
```

Comment:
Nit; I don't think this adds much.

Reply:
Agreed. And with the example moved to ParseTestURL, this doesn't make much sense, either.

```suggestion
```